Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/Model/SingleScalarSqlTableTypeDefinitionProvider.cs

```diff
@@ -0,0 +1,139 @@
+  /// a unique constraint.
+  /// </remarks>
+  public IEnumerable<IRdbmsStructuredTypeDefinition> GetAllStructuredTypeDefinitions ()
+  {
+    yield return GetStructuredTypeDefinition(typeof(string), false);
+    yield return GetStructuredTypeDefinition(typeof(byte[]), false);
+    yield return GetStructuredTypeDefinition(typeof(DummyExtensibleEnum), false);
+
+    yield return GetStructuredTypeDefinition(typeof(bool), false);
+    yield return GetStructuredTypeDefinition(typeof(bool), true);
+    yield return GetStructuredTypeDefinition(typeof(byte), false);
+    yield return GetStructuredTypeDefinition(typeof(byte), true);
+    yield return GetStructuredTypeDefinition(typeof(short), false);
+    yield return GetStructuredTypeDefinition(typeof(short), true);
+    yield return GetStructuredTypeDefinition(typeof(int), false);
+    yield return GetStructuredTypeDefinition(typeof(int), true);
+    yield return GetStructuredTypeDefinition(typeof(long), false);
+    yield return GetStructuredTypeDefinition(typeof(long), true);
+
+    yield return GetStructuredTypeDefinition(typeof(decimal), false);
+    yield return GetStructuredTypeDefinition(typeof(decimal), true);
+    yield return GetStructuredTypeDefinition(typeof(float), false);
+    yield return GetStructuredTypeDefinition(typeof(float), true);
+    yield return GetStructuredTypeDefinition(typeof(double), false);
+    yield return GetStructuredTypeDefinition(typeof(double), true);
+
+    yield return GetStructuredTypeDefinition(typeof(DateTime), false);
+    yield return GetStructuredTypeDefinition(typeof(DateTime), true);
+    yield return GetStructuredTypeDefinition(typeof(Guid), false);
+    yield return GetStructuredTypeDefinition(typeof(Guid), true);
+  }
+
+  /// <summary>
+  /// Gets the <see cref="TableTypeDefinition"/> for the given <paramref name="dotNetType"/> and <paramref name="forDistinctValues"/> flag.
+  /// </summary>
+  /// <param name="dotNetType">The scalar <see cref="Type"/> of the single column.</param>
+  /// <param name="forDistinctValues">Determines whether the <see cref="TableTypeDefinition"/> should have a <see cref="UniqueConstraintDefinition"/> on its single column.</param>
+  /// <remarks>
+  /// A <see cref="UniqueConstraintDefinition"/> is only possible on columns that are up to 900 bytes wide. For example, varchar(max) or varbinary(max) are too wide to have
+  /// a unique constraint.
+  /// </remarks>
+  public IRdbmsStructuredTypeDefinition GetStructuredTypeDefinition (Type dotNetType, bool forDistinctValues)
+  {
+    ArgumentUtility.CheckNotNull(nameof(dotNetType), dotNetType);
+
+    if (typeof(IExtensibleEnum).IsAssignableFrom(dotNetType))
+      dotNetType = typeof(DummyExtensibleEnum); // ExtensibleEnum types have different varchar sizes -> always use our dummy in order to get varchar(max)
+    else
+      dotNetType = Nullable.GetUnderlyingType(dotNetType) ?? dotNetType; // we only have non-nullable columns; ideally, this code should not be needed
```

Comment:
We would get two different `IStorageTypeInformation` objects, one for the nullable type with a nullable column, and one for non-nullable. However, the two `TableTypeDefinition` objects would have the same name, but slightly different `SqlMetaData`. The nullable version would accept null values in AddRecord(), but the command execution would fail because the table type is defined with a non-nullable column.
In contrast, with the current implementation, we should get an `InvalidCastException` during `AddRecord()` when a value is `null`.

Reply:
As discussed, we support nullable value types in the DB.
Don't for get the `GetValue` API which currently has a not-null check.

Reply:
As discussed, we now create table types with nullable columns in the database. We create separate cache items for nullable and non-nullable value types, but both with a nullable `IStorageTypeInformation`.